### ISSUE TITLE
cdist is slow for metric='mahalanobis'

### ISSUE THREAD

**Issue by ForrestCKoch** [scipy/scipy#12570]
I came across https://github.com/scikit-learn/scikit-learn/issues/17813 which suggests calculating Mahalanobis distance as:
d = xT.VI.x + yT.VI.y - xT.VI.y - yT.VI.x
instead of
d = (x-y)T.VI.(x-y)
and so I wrote a quick test to seem how the timings compare cdist with their suggestion and another implementation of Mahalanobis distance.

```python
from time import time

from scipy.spatial.distance import cdist
import numpy as np


SIZE=500
FEATURES=500
x = np.random.normal(size=(FEATURES,SIZE))
y = np.random.normal(size=(FEATURES,SIZE))
vi = np.linalg.inv(np.cov(np.vstack([x.T,y.T]).T)).T

start0 = time()
pwd_scipy = cdist(x.T,y.T,metric='mahalanobis',VI=vi)
fin0 = time() - start0

# calculate (x-y)T.VI.(x-y)
start1 = time()
for i in range(0,SIZE):
    yi =  y[:,i]
    e = (x.T - yi.T).T
    di = np.einsum('ij,ji->i',np.dot(e.T,vi),e)
    assert np.all(np.isclose(pwd_scipy[:,i],np.sqrt(di)))
fin1 = time() - start1


# calculate xT.VI.x + yT.VI.y - xT.VI.y - yT.VI.y
start2 = time()
xtvi = np.dot(x.T,vi)
xtvix = np.einsum('ij,ji->i',xtvi,x)
vix = np.dot(vi,x)
for i in range(0,SIZE):
    yi =  y[:,i]
    ytvi = np.dot(yi.T,vi)
    ytviy = np.dot(ytvi,yi)
    ytvix = np.dot(yi.T,vix)
    xtviy = np.dot(xtvi,yi)
    di = xtvix + ytviy - ytvix - xtviy
    assert np.all(np.isclose(pwd_scipy[:,i],np.sqrt(di)))
fin2 = time() - start2

print('scipy: {}'.format(fin0))
print('Std: {}'.format(fin1))
print('New: {}'.format(fin2))
```
And the output:

```
scipy: 19.262447118759155
Std: 1.1877045631408691
New: 0.05324101448059082
```

What surprised me is that both of my implementations were faster than scipy despite the optimized c code cdist uses.

I'd be happy to open a PR to work on this, but I could use some guidance as to whether it's possible to just plug in the python code or if I need to implement it in C.

My system:
```
>>> sklearn.show_versions()

System:
    python: 3.8.3 (default, May 17 2020, 18:15:42)  [GCC 10.1.0]
executable: /usr/bin/python3
   machine: Linux-5.4.44-1-MANJARO-x86_64-with-glibc2.2.5

Python dependencies:
          pip: 20.0.2
   setuptools: 47.1.1
      sklearn: 0.24.dev0
        numpy: 1.18.1
        scipy: 1.5.1
       Cython: 0.29.21
       pandas: 1.0.1
   matplotlib: 3.2.0
       joblib: 0.14.1
threadpoolctl: 2.1.0

Built with OpenMP: True
```

**Comment by pmla** [comment]
@ForrestCKoch thanks for creating this issue. I have confirmed your timings. This issue is really worth fixing so it is unfortunate that it got overlooked.

@rgommers is there an "owner" of this code? I am happy to help here but this is not an area of `spatial` that I usually work on.

**Comment by rgommers** [comment]
> @rgommers is there an "owner" of this code? I am happy to help here but this is not an area of `spatial` that I usually work on.

Thanks @pmla, help would be much appreciated! `spatial.distance` doesn't have a dedicated maintainer, I think I'm the one who reviews most PRs.

**Comment by peterbell10** [comment]
In general, rearranging distance calculations like can cause accuracy issues. Especially in cases where `x` and `y` are large, but differ on a small scale. For example,
```python
>>> x = 1e12 * np.random.rand(100)
>>> y = x + np.random.rand(100)
>>> ((x-y)**2).sum()
37.50577463819013

>>> (x**2 - 2*x*y + y**2).sum()
-207583232.0
```

Not only is the distance wildly incorrect, it's also negative which should be impossible. So, I'm not sure this is appropriate for a general-purpose library.

As to why cdist is slower than NumPy, I think BLAS is the main reason. The C version implements a naive matrix-vector product.
https://github.com/scipy/scipy/blob/4833a293e7790dd244b2530b74d1a6718cf385d0/scipy/spatial/src/distance_impl.h#L171-L178

That's fine for low dimensional data e.g. for `FEATURES=5` I get SciPy being faster:
```
scipy: 0.02768683433532715
Std: 0.03223299980163574
New: 0.030174732208251953
```

But for 500 dimensions, there's a huge gain from using BLAS. Especially since I think the other versions are all using matrix-matrix BLAS (level 3). 

IMO adding a `gemm`-based version for large dimensional data should make this a lot better.

**Comment by fancidev** [comment]
I happen to come across this issue, and would like to share some idea.

Let $\mathbf X$ be a $p$-by-$n$ matrix with rows denoted by $\mathbf{X}\_i$, $1 \le i \le p$. Similarly, let $\mathbf Y$ be a $q$-by-$n$ matrix with rows denoted by $\mathbf{Y}\_j$, $1 \le j \le q$. Let $\mathbf M$ be an $n$-by-$n$ (inverse covariance) matrix. The goal is to compute matrix $\mathbf D =(d_{ij})$ where $d_{ij}=\sqrt{(\mathbf{X}_i-\mathbf{Y}_j)\mathbf{M}  (\mathbf{X}_i-\mathbf{Y}_j)’}$.

**Comment by fancidev** [comment]
If the inverse covariance matrix is not given but estimated from the input, then the computations are similar except that a little saving is possible by omitting the inversion and Cholesky decomposition. This seems to have some traction as discussed [here](https://github.com/scipy/scipy/issues/4815#issuecomment-99231104); see also gh-4815 and gh-4886.

Suppose the covariance is to be estimated from an $m$-by-$n$ matrix $\mathbf Z$ where each row contains one observation. Suppose $m > n$, so that the sample covariance matrix is not singular. Suppose the mean of each column of $\mathbf Z$ is zero; if they are not, they can be centered to be so. The sample covariance matrix is $\mathbf C=\frac{1}{m-1} \mathbf Z’ \mathbf Z$.

**Comment by fancidev** [comment]
If a nonnegative weight scalar $w$ or vector $\mathbf w$ is supplied, the “weighted” Mahalanobis distance may be defined as

$$
\sqrt{ (\mathbf X_i-\mathbf Y_j) \mathbf W^{\frac 1 2}\mathbf M \mathbf W^{\frac 1 2}(\mathbf X_i-\mathbf Y_j)’}
$$

where $\mathbf W=w \mathbf I$ or $\mathrm{diag}(\mathbf w)$. This is just a matter of left-multiplying $\mathbf L$ or $\mathbf U$ by $\mathbf W^{\frac 1 2}$ and the computations proceed as usual.

**Comment by ilayn** [comment]
I know this is adding not much to what you show here but `vec` x `mat` x `vec` operation in the middle in the current code is significantly slower as @peterbell10 mentioned above. For large arrays this is quite a difference at the sizes you mention that changes your constant in your big O notation.

Grouping this `vec` x (`mat` x `vec`) with two `gemv` calls is going to be your baseline in case you would like to start from an improvement. Then you can decide whether you want to include a `potrf` and a `trtrs` instead of a `gemv` call.

So I would start from there even in numpy code you can get that baseline.
